nltk/sentiment/vader.py

```diff
@@ -433,30 +448,76 @@ def _least_check(self, valence, words_and_emoticons, i):
             i > 0
             and words_and_emoticons[i - 1].lower() not in self.lexicon
             and words_and_emoticons[i - 1].lower() == "least"
         ):
-            valence = valence * N_SCALAR
+            valence = valence * self.constants.N_SCALAR
         return valence
 
-    def _but_check(self, words_and_emoticons, sentiments):
-        # check for modification in sentiment due to contrastive conjunction 'but'
-        if "but" in words_and_emoticons or "BUT" in words_and_emoticons:
-            try:
-                bi = words_and_emoticons.index("but")
-            except ValueError:
-                bi = words_and_emoticons.index("BUT")
-            for sentiment in sentiments:
-                si = sentiments.index(sentiment)
-                if si < bi:
-                    sentiments.pop(si)
-                    sentiments.insert(si, sentiment * 0.5)
-                elif si > bi:
-                    sentiments.pop(si)
-                    sentiments.insert(si, sentiment * 1.5)
+    def _but_check(self, words_and_emoticons, sentiments, extended=False):
+        if extended:
+            # Extended version from
+            # https://medium.com/@malavika.suresh0794/vader-customizing-the-library-71d9e8ed6eda
+            # check for modification in sentiment due to contrastive conjunctions
+            words_and_emoticons_lower = [str(w).lower() for w in words_and_emoticons]
+            cc_list = ['but', 'however', 'except']
+            bi = 0
+            for cc in cc_list:
+                if cc in words_and_emoticons_lower:
+                    bi = words_and_emoticons_lower.index(cc)
+                    for si,sentiment in enumerate(sentiments):
+                        if si < bi:
+                            sentiments[si] = sentiment * 0.5
+                        elif si > bi:
+                            sentiments[si] = sentiment * 1.5
+            # Future work:
+            # 1.Consider usage of though/although/even though
+            return sentiments
+        else:
+            # Original Vader code.
+            # check for modification in sentiment due to contrastive conjunction 'but'
+            if "but" in words_and_emoticons or "BUT" in words_and_emoticons:
+                try:
+                    bi = words_and_emoticons.index("but")
+                except ValueError:
+                    bi = words_and_emoticons.index("BUT")
+                for sentiment in sentiments:
+                    si = sentiments.index(sentiment)
+                    if si < bi:
+                        sentiments.pop(si)
+                        sentiments.insert(si, sentiment * 0.5)
+                    elif si > bi:
+                        sentiments.pop(si)
+                        sentiments.insert(si, sentiment * 1.5)
+            return sentiments
+
+    def _only_if_check(self, words_and_emoticons, sentiments):
+        words_and_emoticons_lower = [str(w).lower() for w in words_and_emoticons]
```

Comment:
Would `words_and_emoticons` change in the process of the algorithm, if it doesn't putting it as a class variable and accessing it might be more efficient.

Reply:
I think the suggestion makes sense, I will look to incorporate it and also benchmark against a fixed test set. Really appreciate the comments and efforts !

Reply:
Yes, thanks @alvations